// @sudoku/stores/solver.js
class SudokuParser {
    constructor() {
        this.baseUrl = 'https://www.sudokuwiki.org/sudoku.htm';
    }

    /**
     * 解析URL中的bd参数，获取数独字符串
     * @param {string} url - 完整URL
     * @returns {string|null} 81位数独字符串
     */
    parseUrl(url) {
        try {
            const urlObj = new URL(url);
            const bdParam = urlObj.searchParams.get('bd');
            
            if (!bdParam) {
                throw new Error('URL中未找到bd参数');
            }
            
            return this.parseBdParameter(bdParam);
        } catch (error) {
            console.error('URL解析失败:', error);
            return null;
        }
    }

    /**
     * 解析bd参数为81位字符串
     * @param {string} bdParam - bd参数值
     * @returns {string} 81位字符串，空白用0表示
     */
    parseBdParameter(bdParam) {
        // 检查参数长度
        if (bdParam.length !== 81) {
            console.warn(`bd参数长度异常: ${bdParam.length}，尝试清理...`);
        }
        
        // 清理非数字字符，并将0-9以外的字符转换为0
        let cleaned = '';
        for (let char of bdParam) {
            if (char >= '0' && char <= '9') {
                cleaned += char;
            } else {
                cleaned += '0'; // 非数字字符视为空白
            }
        }
        
        // 如果长度不足81，用0补齐
        if (cleaned.length < 81) {
            cleaned = cleaned.padEnd(81, '0');
        }
        
        // 如果长度超过81，截取前81位
        if (cleaned.length > 81) {
            cleaned = cleaned.substring(0, 81);
        }
        
        return cleaned;
    }

    /**
     * 验证数独字符串
     * @param {string} sudokuString 
     * @returns {boolean}
     */
    validate(sudokuString) {
        if (!sudokuString || typeof sudokuString !== 'string') {
            return false;
        }
        
        if (sudokuString.length !== 81) {
            return false;
        }
        
        // 检查是否只包含0-9
        return /^[0-9]{81}$/.test(sudokuString);
    }

    /**
     * 格式化显示数独
     * @param {string} sudokuString 
     * @returns {string}
     */
    format(sudokuString) {
        if (!this.validate(sudokuString)) {
            return '无效的数独字符串';
        }
        
        let formatted = '';
        for (let i = 0; i < 9; i++) {
            const row = sudokuString.substr(i * 9, 9);
            const displayRow = row.replace(/0/g, '.');
            
            // 每3个数字加空格
            const parts = [];
            for (let j = 0; j < 9; j += 3) {
                parts.push(displayRow.substr(j, 3));
            }
            formatted += parts.join(' ');
            
            // 添加行分隔符
            if (i < 8) {
                formatted += '\n';
                if (i === 2 || i === 5) {
                    formatted += '------+-------+------\n';
                }
            }
        }
        return formatted;
    }


}

const solver = new SudokuParser()

export default solver
